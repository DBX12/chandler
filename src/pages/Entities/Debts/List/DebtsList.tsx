import React, {PureComponent} from 'react';
import {ActivityIndicator, FlatList, Text, View} from 'react-native';

import {EmptyActivity, Navbar} from '@components';
import {I18n} from '@i18n';
import {IDebt} from '@models';
import {IPopAction} from '@navigator/NavigationService';
import {commonStyles} from '@theme';
import {styles} from './DebtsList.styles';

interface IDebtsListProps {
  pop: IPopAction;
  getDebtsByContact: () => void;
  debts: IDebt[];
  isFetching: boolean;
}

export class DebtsList extends PureComponent<IDebtsListProps, {}> {
  public componentWillMount() {
    this.props.getDebtsByContact();
  }

  public keyExtractor = (item, index) => String(item.id);

  public renderFooter = () => {
    const {isFetching} = this.props;

    if (!isFetching) {
      return null;
    }

    return (
      <ActivityIndicator style={commonStyles.activityIndicator} size="large" />
    );
  };

  public renderItem = ({index}) => {
    const {debts} = this.props;
    const debt = debts[index];

    return (
      <View style={styles.debtContainer}>
        <View style={styles.headerContainer}>
          <View
            style={[
              styles.badge,
              debt.in_debt === 'no' ? styles.badgeSuccess : styles.badgeError,
            ]}
          >
            {debt.in_debt === 'yes' && (
              <Text style={styles.badgeText}>{I18n.t('debts:youOwe')}</Text>
            )}
            {debt.in_debt === 'no' && (
              <Text style={styles.badgeText}>
                {I18n.t('debts:owesYou', {name: debt.contact.first_name})}
              </Text>
            )}
          </View>

          <Text style={styles.amountText}>${debt.amount}</Text>
        </View>
        <Text style={styles.reasonText}>{debt.reason}</Text>
      </View>
    );
  };

  public render() {
    const {pop, debts, getDebtsByContact, isFetching} = this.props;

    return (
      <View style={commonStyles.flex}>
        <Navbar title={I18n.t('debts:debts')} onBack={pop} />
        {isFetching || debts.length ? (
          <FlatList
            data={debts}
            renderItem={this.renderItem}
            keyExtractor={this.keyExtractor}
            ListFooterComponent={this.renderFooter}
            onEndReached={getDebtsByContact}
            onEndReachedThreshold={0.5}
          />
        ) : (
          <EmptyActivity
            image={require('./assets/empty-debts.png')}
            title={I18n.t('debts:emptyTitle')}
            subtitle={I18n.t('debts:emptySubtitle')}
          />
        )}
      </View>
    );
  }
}
