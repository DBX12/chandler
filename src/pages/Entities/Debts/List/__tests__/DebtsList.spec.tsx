import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import {DebtsList} from '../DebtsList';

describe('Pages', () => {
  describe('DebtsList', () => {
    let pop;
    let getDebtsByContact;

    beforeEach(() => {
      pop = jest.fn();
      getDebtsByContact = jest.fn();
    });

    it('should renders correctly without debts and fetching', () => {
      const debts = [];
      const tree = shallow(
        <DebtsList
          pop={pop}
          getDebtsByContact={getDebtsByContact}
          isFetching={true}
          debts={debts}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
    });

    it('should renders correctly without debts and not fetching', () => {
      const debts = [];
      const tree = shallow(
        <DebtsList
          pop={pop}
          getDebtsByContact={getDebtsByContact}
          isFetching={false}
          debts={debts}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
    });

    it('should renders correctly with debts and fetching', () => {
      const debts = [
        {in_debt: 'yes', reason: 'My reason', contact: {first_name: 'Theo'}},
      ];
      const tree = shallow(
        <DebtsList
          pop={pop}
          getDebtsByContact={getDebtsByContact}
          isFetching={true}
          debts={debts as any}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 0})))
      ).toMatchSnapshot();
    });

    it('should renders correctly with debts and fetching', () => {
      const debts = [
        {in_debt: 'yes', reason: 'My reason', contact: {first_name: 'Theo'}},
        {in_debt: 'no', reason: 'My reason 2', contact: {first_name: 'Bill'}},
      ];
      const tree = shallow(
        <DebtsList
          pop={pop}
          getDebtsByContact={getDebtsByContact}
          isFetching={false}
          debts={debts as any}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 0})))
      ).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 1})))
      ).toMatchSnapshot();
    });

    it('should display the activity indicator in the footer if it is fetching', () => {
      const debts = [
        {body: 'My body', created_at: '1991-12-17'},
        {body: 'My body 2', created_at: '1993-12-17'},
      ];
      const tree = shallow(
        <DebtsList
          pop={pop}
          getDebtsByContact={getDebtsByContact}
          isFetching={true}
          debts={debts as any}
        />
      );
      expect(toJson(shallow(tree.instance().renderFooter()))).toMatchSnapshot();
    });

    it('should not display the activity indicator in the footer if it is not fetching', () => {
      const debts = [
        {body: 'My body', created_at: '1991-12-17'},
        {body: 'My body 2', created_at: '1993-12-17'},
      ];
      const tree = shallow(
        <DebtsList
          pop={pop}
          getDebtsByContact={getDebtsByContact}
          isFetching={false}
          debts={debts as any}
        />
      );
      expect(tree.instance().renderFooter()).toBeNull();
    });

    it('should have a keyExtractor', () => {
      const debts = [{id: 1}, {id: 2}];
      const tree = shallow(
        <DebtsList
          pop={pop}
          getDebtsByContact={getDebtsByContact}
          isFetching={false}
          debts={debts as any}
        />
      );
      expect(tree.instance().keyExtractor(debts[0])).toBe('1');
      expect(tree.instance().keyExtractor(debts[1])).toBe('2');
    });
  });
});
