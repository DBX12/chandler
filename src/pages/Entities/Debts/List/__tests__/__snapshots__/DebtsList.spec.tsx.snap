// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages DebtsList should display the activity indicator in the footer if it is fetching 1`] = `
<ActivityIndicator
  animating={true}
  color="#999999"
  hidesWhenStopped={true}
  size="large"
  style={
    Object {
      "marginVertical": 20,
    }
  }
/>
`;

exports[`Pages DebtsList should renders correctly with debts and fetching 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Debts"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={
      Array [
        Object {
          "contact": Object {
            "first_name": "Theo",
          },
          "in_debt": "yes",
          "reason": "My reason",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Component>
`;

exports[`Pages DebtsList should renders correctly with debts and fetching 2`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <Component
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 10,
      }
    }
  >
    <Component
      style={
        Array [
          Object {
            "borderRadius": 5,
            "paddingHorizontal": 5,
            "paddingVertical": 2,
          },
          Object {
            "backgroundColor": "red",
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontSize": 13,
          }
        }
      >
        You owe
      </Text>
    </Component>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "marginLeft": 6,
          "marginTop": 2,
        }
      }
    >
      $
    </Text>
  </Component>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 13,
      }
    }
  >
    My reason
  </Text>
</View>
`;

exports[`Pages DebtsList should renders correctly with debts and fetching 3`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Debts"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={
      Array [
        Object {
          "contact": Object {
            "first_name": "Theo",
          },
          "in_debt": "yes",
          "reason": "My reason",
        },
        Object {
          "contact": Object {
            "first_name": "Bill",
          },
          "in_debt": "no",
          "reason": "My reason 2",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Component>
`;

exports[`Pages DebtsList should renders correctly with debts and fetching 4`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <Component
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 10,
      }
    }
  >
    <Component
      style={
        Array [
          Object {
            "borderRadius": 5,
            "paddingHorizontal": 5,
            "paddingVertical": 2,
          },
          Object {
            "backgroundColor": "red",
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontSize": 13,
          }
        }
      >
        You owe
      </Text>
    </Component>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "marginLeft": 6,
          "marginTop": 2,
        }
      }
    >
      $
    </Text>
  </Component>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 13,
      }
    }
  >
    My reason
  </Text>
</View>
`;

exports[`Pages DebtsList should renders correctly with debts and fetching 5`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <Component
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 10,
      }
    }
  >
    <Component
      style={
        Array [
          Object {
            "borderRadius": 5,
            "paddingHorizontal": 5,
            "paddingVertical": 2,
          },
          Object {
            "backgroundColor": "#355974",
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontSize": 13,
          }
        }
      >
        Bill owes you
      </Text>
    </Component>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "marginLeft": 6,
          "marginTop": 2,
        }
      }
    >
      $
    </Text>
  </Component>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 13,
      }
    }
  >
    My reason 2
  </Text>
</View>
`;

exports[`Pages DebtsList should renders correctly without debts and fetching 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Debts"
  />
  <FlatList
    ListFooterComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Component>
`;

exports[`Pages DebtsList should renders correctly without debts and not fetching 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Debts"
  />
  <EmptyActivity
    image={1}
    subtitle="But having no debts that you owe is a way to stay happy in your friendship."
    title="Money doesnâ€™t buy happiness."
  />
</Component>
`;
